docs/source/miscellaneous/doc/adr/0014-cassandra-targeted-sweep.md

```diff
@@ -0,0 +1,283 @@
+# 14. Cassandra Targeted Sweep
+
+Date: 18/06/2018
+
+## Status
+
+Accepted
+
+## Context
+
+### Legacy Sweep
+
+To achieve its transactional guarantees, AtlasDB maintains historical versions of cells that have been written to. Eventually,
+all existing and future transactions will have a start timestamp large enough that some of these historical versions will never be
+visible again, resulting in unnecessary cruft. This is not only an issue due to taking up storage space in the underlying KVS, but
+also because certain access patterns require scanning over the obsolete historic versions leading to significant performance
+degradation over time. The process of removing old historic versions of cells from tables in AtlasDB is called **Sweep**.
+
+We will refer to the current implementation of sweep that AtlasDB is relying on as **Legacy Sweep**. Legacy sweep is an iterative
+procedure that, given a table reference and a start row, sequentially scans through all the historic versions of each cell searching
+for candidates written in transactions with a start timestamp and a commit timestamp both lower than the **Sweep Timestamp**. The
+sweep timestamp is a timestamp expected to be lower than the start timestamp of all open transactions, rendering all but the last
+historic version for each cell prior to the sweep timestamp effectively invisible and thus obsolete. These versions can therefore be
+safely deleted.
+
+Legacy sweep continues these scans until enough candidates have been found, processing at least one full row of cells, and
+then deletes each of the obsolete historic versions. There are two main modes for running legacy sweep:
+
+1. Background Sweep, which is a background task that repeatedly chooses a table to sweep and then proceeds to run iterations
+of sweep until all the rows have been processed.
+2. Manual Sweep, which can be triggered through a REST endpoint to perform an iteration or a full sweep for a given table
+and start row.
+
+Note that there are corner cases where the sweep timestamp may be mistaken: a read only transaction that has been running
+for longer than one hour (which is also subject to clock drift), or when a write transaction loses its lock while committing. To
+ensure correctness for read only transactions with respect to these cases, on tables that allow read only transactions we
+defensively write a **garbage deletion sentinel**, which is an empty value at a timestamp of -1, i.e., before any valid start
+timestamp. If a read only transactions encounters a sentinel, that signalizes there could have been a historic version it should be
```

Comment:
Technically, we can have a long running read-only transaction, a first version of the cell committed after the read-only transaction has started, and then swept away. This would leave a sentinel, but there was nothing for the read-only transaction to see there